torch/csrc/autograd/variable.h

```diff
@@ -522,6 +524,8 @@ inline Variable make_variable(
     at::Tensor data,
     Edge gradient_edge,
     bool allow_tensor_metadata_change = true) {
+  TORCH_CHECK(isDifferentiableType(at::typeMetaToScalarType(data.dtype())),
```

Comment:
why is this one necessary?  Isn't the code below (line 533) creating a variable with requires_grad set to false?

Reply:
The semantic of `requires_grad_` in autograd meta is not the same as the one in python here.
The `requires_grad` flag is only relevant for leaf Tensors.
So a tensore requires grad is it has `requires_grad_` or has a grad_fn (code [here](https://github.com/pytorch/pytorch/blob/7ae41d5d329a65b200dbd448794ce228b10bc901/torch/csrc/autograd/variable.h#L228-L230)).

So this one is implicitly creating a Tensor that requires grad from the python definition.